File: library/foundation/android/java/src/library/foundation/BRIDGE.java
package src.library.foundation;

@SuppressWarnings("unused") /* the methods will be called by native */
public class BRIDGE {

    //log:

    public static void cq_log_info(String file, int line, String message) {
        L.i("%s", message);
    }

    public static void cq_log_error(String file, int line, String message) {
        L.e("%s", message);
    }

    //file manager:

    public static String cq_document_directory() {
        return FileManager.documentDirectory();
    }

    public static String cq_caches_directory() {
        return FileManager.cachesDirectory();
    }

    public static String cq_temporary_directory() {
        return FileManager.temporaryDirectory();
    }

    public static String cq_append_path(String parent, String child) {
        return FileManager.appendPath(parent, child);
    }

    public static boolean cq_directory_exists(String path) {
        return FileManager.get().directoryExists(path);
    }

    public static boolean cq_file_exists(String path) {
        return FileManager.get().fileExists(path);
    }

    public static boolean cq_create_directory(String path, boolean intermediate) {
        return FileManager.get().createDirectory(path, intermediate);
    }

    public static void cq_remove_path(String path) {
        FileManager.get().removePath(path);
    }

    //thread:

    public static void cq_thread_run(long task, long data) {
        ThreadManager.run(() -> threadBody(task, data));
    }

    private static native void threadBody(long task, long data);

    public static void cq_thread_sleep(float seconds) {
        ThreadManager.sleepForSeconds(seconds);
    }

    //network:

    public static void cq_http_get(String url, float timeout) {
        int[] error = new int[1];
        byte[] bytes = URLSession.get().sendSyncGet(url, timeout, error);
        httpGetReturn(error[0], bytes);
    }

    public static native void httpGetReturn(int error, byte[] data);
}

File: library/foundation/win32foundation/BRIDGE.c
#include "cqfoundationarche.h"
#include "cqwin32foundation.h"

//log:

void cq_log_info(const char *file, int32_t line, const char *message)
{
    cq_alloc_pool
    {
        LPCWSTR pszFile = cq_alloc_wstr(cq_wsfrom8s(file));
        LPCWSTR pszMessage = cq_alloc_wstr(cq_wsfrom8s(message));

        CQLogInfoW(pszFile, line, pszMessage);
    }
    cq_free_pool();
}

void cq_log_error(const char *file, int32_t line, const char *message)
{
    cq_alloc_pool
    {
        LPCWSTR pszFile = cq_alloc_wstr(cq_wsfrom8s(file));
        LPCWSTR pszMessage = cq_alloc_wstr(cq_wsfrom8s(message));

        CQLogErrorW(pszFile, line, pszMessage);
    }
    cq_free_pool();
}

//file management:

const char *cq_document_directory(void)
{
    LPCWSTR pszPath = CQDocumentDirectoryW();
    return cq_u8sfromws(pszPath);
}

const char *cq_caches_directory(void)
{
    LPCWSTR pszPath = CQCachesDirectoryW();
    return cq_u8sfromws(pszPath);
}

const char *cq_temporary_directory(void)
{
    LPCWSTR pszPath = CQTemporaryDirectoryW();
    return cq_u8sfromws(pszPath);
}

bool cq_directory_exists(const char *path)
{
    LPCWSTR pszPath = cq_wsfrom8s(path);
    return CQDirectoryExistsW(pszPath);
}

bool cq_file_exists(const char *path)
{
    LPCWSTR pszPath = cq_wsfrom8s(path);
    return CQFileExistsW(pszPath);
}

bool cq_create_directory(const char *path, bool intermediate)
{
    LPCWSTR pszPath = cq_wsfrom8s(path);
    return CQCreateDirectoryW(pszPath, intermediate);
}

void cq_remove_path(const char *path)
{
    LPCWSTR pszPath = cq_wsfrom8s(path);
    CQRemovePathW(pszPath);
}

//thread:

void cq_thread_run(void (*task)(void *), void *data)
{
    CQRunThread(task, data);
}

void cq_thread_sleep(float seconds)
{
    CQThreadSleepSeconds(seconds);
}
